src/android.c

```diff
@@ -0,0 +1,402 @@
+        outstream->bytes_per_sample * 8,
+        outstream->layout.channel_count == 1 ? SL_SPEAKER_FRONT_CENTER :
+            (SL_SPEAKER_FRONT_LEFT | SL_SPEAKER_FRONT_RIGHT),
+        SL_BYTEORDER_LITTLEENDIAN,
+        outstream->format == SoundIoFormatFloat32LE ?
+            SL_ANDROID_PCM_REPRESENTATION_FLOAT :
+            outstream->format == SoundIoFormatS16LE ?
+            SL_ANDROID_PCM_REPRESENTATION_SIGNED_INT :
+            SL_ANDROID_PCM_REPRESENTATION_UNSIGNED_INT
+    };
+
+    SLDataSource audioSrc = {&loc_bufq, &format_pcm};
+    SLDataLocator_OutputMix loc_outmix = {
+        SL_DATALOCATOR_OUTPUTMIX,
+        osa->outputMixObject
+    };
+    SLDataSink audioSnk = {&loc_outmix, NULL};
+    SLInterfaceID ids[1] = {SL_IID_BUFFERQUEUE};
+    const SLboolean req[1] = {SL_BOOLEAN_TRUE};
+
+    if (SL_RESULT_SUCCESS != (*sia->engineEngine)->CreateAudioPlayer(sia->engineEngine,
+        &osa->playerObject, &audioSrc, &audioSnk, 1, ids, req))
+    {
+        outstream_destroy_android(si, os);
+        return SoundIoErrorOpeningDevice;
+    }
+    if (SL_RESULT_SUCCESS != (*osa->playerObject)->Realize(osa->playerObject,
+        SL_BOOLEAN_FALSE))
+    {
+        outstream_destroy_android(si, os);
+        return SoundIoErrorOpeningDevice;
+    }
+
+    if (SL_RESULT_SUCCESS != (*osa->playerObject)->GetInterface(osa->playerObject,
+        SL_IID_PLAY, &osa->playerPlay))
+    {
+        outstream_destroy_android(si, os);
+        return SoundIoErrorOpeningDevice;
+    }
+
+    if (SL_RESULT_SUCCESS != (*osa->playerObject)->GetInterface(osa->playerObject,
+        SL_IID_BUFFERQUEUE, &osa->playerBufferQueue))
+    {
+        outstream_destroy_android(si, os);
+        return SoundIoErrorOpeningDevice;
+    }
+    if (SL_RESULT_SUCCESS != (*osa->playerBufferQueue)->RegisterCallback(
+        osa->playerBufferQueue, write_callback_android, os))
+    {
+        outstream_destroy_android(si, os);
+        return SoundIoErrorOpeningDevice;
+    }
+
+    // TODO: An audio stream's latency and jitter can be reduced by using the
+    // native buffer size. Unfortunately, there is no officially supported way
+    // to get the native buffer size without access to a JVM.
+    osa->bytes_per_buffer = outstream->bytes_per_frame * outstream->sample_rate
+        * outstream->software_latency / 2;
+    size_t samples_per_buffer = osa->bytes_per_buffer / outstream->bytes_per_sample;
+    osa->buffers[0] = calloc(samples_per_buffer, outstream->bytes_per_sample);
+    osa->buffers[1] = calloc(samples_per_buffer, outstream->bytes_per_sample);
+
+    return 0;
+}
+
+static int outstream_pause_android(struct SoundIoPrivate *si,
+    struct SoundIoOutStreamPrivate *os, bool pause)
+{
+    struct SoundIoOutStreamAndroid *osa = &os->backend_data.android;
+
+    if (SL_RESULT_SUCCESS != (*osa->playerPlay)->SetPlayState(osa->playerPlay,
+        pause ? SL_PLAYSTATE_PAUSED : SL_PLAYSTATE_PLAYING))
+    {
+        return SoundIoErrorInvalid;
```

Comment:
`SoundIoErrorInvalid` I believe is meant to mean "you called this function with invalid state" such as passing NULL for `si`. If SetPlayState returns an error, I think we want one of these:

```
    /// An open stream had an error that can only be recovered from by
    /// destroying the stream and creating it again.
    SoundIoErrorStreaming,

    /// Attempted to use a device with parameters it cannot support.
    SoundIoErrorIncompatibleDevice,

    /// Attempted to use parameters that the backend cannot support.
    SoundIoErrorIncompatibleBackend,
```